Decompress the previous sibling and renderer pointers in WebCore::Node

ddegazio opened the pull request:
#### 113f84cbe25484e4cb4f167259c65e6200b651be
<pre>
Decompress the previous sibling and renderer pointers in WebCore::Node
<a href="https://bugs.webkit.org/show_bug.cgi?id=281997">https://bugs.webkit.org/show_bug.cgi?id=281997</a>
<a href="https://rdar.apple.com/138409833">rdar://138409833</a>

Reviewed by Ryosuke Niwa.

Makes use of some currently unused padding bytes in WebCore::Node to
decompress two CompactPointerTuple fields, simplifying access to these
fields without increasing the memory footprint of Node itself.

* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::SelectorCodeGenerator::generateWalkToPreviousAdjacentElement):
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::~Node):
* Source/WebCore/dom/Node.h:
(WebCore::Node::previousSibling const):
(WebCore::Node::protectedPreviousSibling const):
(WebCore::Node::previousSiblingMemoryOffset):
(WebCore::Node::setPreviousSibling):
(WebCore::Node::renderer const):
(WebCore::Node::hasElementStateFlag const):
(WebCore::Node::styleBitfields const):
(WebCore::Node::setStyleBitfields):
(WebCore::Node::setElementStateFlag const):
(WebCore::Node::previousSiblingPointerMask): Deleted.
* Source/WebCore/dom/NodeRareData.cpp:
* Source/WebCore/domjit/JSNodeDOMJIT.cpp:
(WebCore::compileNodePreviousSiblingAttribute):
* Source/WebCore/rendering/RenderObject.cpp:
* Source/WebCore/rendering/RenderObject.h:
(WebCore::Node::setRenderer):

Canonical link: <a href="https://commits.webkit.org/286827@main">https://commits.webkit.org/286827@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/35bd9228c67856b9969bb05dd3a927c6a909dd5b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77209 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56244 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/30124 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81772 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/28491 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79326 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65392 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4540 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/60494 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/28491 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80276 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50455 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66271 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/40794 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47858 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23769 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/26814 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/68970 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/24104 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83197 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4589 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/3094 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/68773 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4745 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66244 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/68030 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16976 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/12009 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/10106 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4536 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/7351 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/4555 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/7990 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6314 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -118,11 +118,13 @@ struct SameSizeAsNode : EventTarget, CanMakeCheckedPtr<SameSizeAsNode> {
     void* parentNode;
     void* treeScope;
-    uint8_t previous[8];
+    void* previous;
```
Does this have to be a `void*`?  Can it be `Node*`?

ddegazio replied:
This type is just a proxy for the layout of `Node` so we can `static_assert` that `Node` doesn't unexpectedly change in size. I'm just updating it to reflect the types a bit better, but there's no need for it to be more precise so long as the size and alignment of fields are the same.

MenloDorian commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -118,11 +118,13 @@ struct SameSizeAsNode : EventTarget, CanMakeCheckedPtr<SameSizeAsNode> {
+    void* previous;
     void* next;
-    uint8_t rendererWithStyleFlags[8];
+    void* renderer;
```
Ditto: can this be `RenderObject*` instead?

ddegazio replied:
See my previous comment.

MenloDorian commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -118,11 +118,13 @@ struct SameSizeAsNode : EventTarget, CanMakeCheckedPtr<SameSizeAsNode> {
 #endif
     uint32_t refCountAndParentBit;
     uint32_t nodeFlags;
+    uint16_t elementStateFlags;
+    uint16_t styleBitfields;
```
I don't see these used anywhere else.  Are they needed?

ddegazio replied:
See previous comment - these specifically reflect the `OptionSet<ElementStateFlag>` and `StyleBitfields` unpacked from the previous compact tuples in `Node`.

kmiller68 commented on Source/WebCore/dom/Node.h:
```diff
@@ -797,12 +790,14 @@ class Node : public EventTarget, public CanMakeCheckedPtr<Node> {
     CheckedPtr<ContainerNode> m_parentNode;
     TreeScope* m_treeScope { nullptr };
-    CompactPointerTuple<Node*, uint16_t> m_previous;
+    Node* m_previousSibling;
```
Shouldn't this be a CheckedPtr?

cdumez replied:
At the very least: `{ nullptr };`? Who is default initializing this otherwise?

ddegazio replied:
Definitely these are missing a `{ nullptr }`, good catch - I've added null initialization in the latest push. On changing these to `CheckedPtr`, we discussed this elsewhere and that might be more suitable for a different future patch due to the potential performance impact.

kmiller68 commented on Source/WebCore/dom/Node.h:
```diff
@@ -797,12 +790,14 @@ class Node : public EventTarget, public CanMakeCheckedPtr<Node> {
+    Node* m_previousSibling;
     CheckedPtr<Node> m_next;
-    CompactPointerTuple<RenderObject*, uint16_t> m_rendererWithStyleFlags;
+    RenderObject* m_renderer;
```
Ditto.

ddegazio replied:
This got a `{ nullptr }` in the latest push as well.

ddegazio committed:
Decompress the previous sibling and renderer pointers in WebCore::Node
https://bugs.webkit.org/show_bug.cgi?id=281997
rdar://138409833

Reviewed by Ryosuke Niwa.

Makes use of some currently unused padding bytes in WebCore::Node to
decompress two CompactPointerTuple fields, simplifying access to these
fields without increasing the memory footprint of Node itself.

* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::SelectorCodeGenerator::generateWalkToPreviousAdjacentElement):
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::~Node):
* Source/WebCore/dom/Node.h:
(WebCore::Node::previousSibling const):
(WebCore::Node::protectedPreviousSibling const):
(WebCore::Node::previousSiblingMemoryOffset):
(WebCore::Node::setPreviousSibling):
(WebCore::Node::renderer const):
(WebCore::Node::hasElementStateFlag const):
(WebCore::Node::styleBitfields const):
(WebCore::Node::setStyleBitfields):
(WebCore::Node::setElementStateFlag const):
(WebCore::Node::previousSiblingPointerMask): Deleted.
* Source/WebCore/dom/NodeRareData.cpp:
* Source/WebCore/domjit/JSNodeDOMJIT.cpp:
(WebCore::compileNodePreviousSiblingAttribute):
* Source/WebCore/rendering/RenderObject.cpp:
* Source/WebCore/rendering/RenderObject.h:
(WebCore::Node::setRenderer):

Canonical link: https://commits.webkit.org/286827@main
